storage/innobase/row/row0ins.cc

```diff
@@ -2636,9 +2636,15 @@ static dberr_t row_ins_sorted_clust_index_entry(ulint mode, dict_index_t *index,
   mtr = &index->last_ins_cur->mtr;
 
   /* Search for position if tree needs to be split or if last position
-  is not cached. */
+  is not cached. Also commit mtr and do the search if we have inserted
+  more than MAX_INTRINSIC_MTR_RECORDS within this mtr. The latter is
+  done to unfix buffer pages used by the mtr. Otherwise statements
+  (even single statement in extreme case) using intrinsic tables can
+  consume the whole buffer pool by their buffer fixed pages causing
+  performance problems or even stalls. */
   if (mode == BTR_MODIFY_TREE || index->last_ins_cur->rec == nullptr ||
-      index->last_ins_cur->disable_caching) {
+      index->last_ins_cur->disable_caching ||
+      index->last_ins_cur->mtr_records > MAX_INTRINSIC_MTR_RECORDS) {
```

Comment:
Can we use this new condition and set index->last_ins_cur->disable_caching?

Reply:
Hello Satya!

In theory we can. But I think we should not do this due to following reasons:

1) Currently semantics of disable_caching assumes that it stays active even after mtr-commit (until end
    of statement?)... It indicates condition related to table/query structure rather than something specific
    to the row being processed. If we are to change this we will have to add code which will clear this flag
    at least in some cases on mtr-commit and as result we will end up in more complex code.
2) code in row0sel.cc doesn't check this condition, so either we will get discrepancy between insert
    and select cases, or we will have to update row0sel.cc as well (and code for its clearing on mtr-commit
    too). Again the code will become more complex.

What do you think?

Reply:
Ok, the intention of the suggestion is to make it simpler :) OK to keep it as is.